### ISSUE TITLE
Swapping to remote branches

### ISSUE THREAD

**Issue by mikemicmike** [microsoft/vscode#88983]
Version: 1.41.1 (system setup)
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T14:58:56.166Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.18362

**Steps to Reproduce:**
No need for steps really the bahaviour speaks for itself. 
When swapping branches, even selecting the "Origin/Master" doesn't seem to pickup that commit number properly. 
![Branches](https://user-images.githubusercontent.com/17456036/72745210-3ab34780-3b7d-11ea-905f-a89297db76b5.png)
So in that screenshot you can see there are 2 "masters" the local and the remote (Origin/Master). 
The remote is 6 commits ahead. When i checkout that remote branch i expect to be at the latest commit not to be at the same commit as the local branch. 
![After swap to remote master](https://user-images.githubusercontent.com/17456036/72745291-6b937c80-3b7d-11ea-964d-9f037d3c26b7.png)

There seems to be no difference no matter what i swap to when in reality swapping to a remote branch should pull those immediately as you've already specified you are pulling a remote branch. (Thats my understanding anyways but if it isn't the case then there is no use in bothering to show both versions of the branch.) 

Thanks! 

Does this issue occur when all extensions are disabled?: Yes

**Comment by joaomoreno** [comment]
Interesting. When you checkout to a remote branch we automatically create a local branch which tracks it. If a local branch already tracks it, we choose that instead. That's what's happening: your local `master` branch already tracks `origin/master`, so it gets picked.

**Comment by mikemicmike** [comment]
What do you mean would you prefer to checkout headless? 

I think in reality what should happen is that if you switch branch to a remote branch, its fine to swap to the local branch but an automatic pull should be run straight after swapping to a remote branch. If you pick the local branch then no automatic pull.

**Comment by joaomoreno** [comment]
You can't checkout to a remote branch, the concept doesn't exist in Git. There are a few options of what's possible:

1. Checkout to a local branch which tracks the remote branch, if existent :ok_hand: this is what we currently do
2. Checkout to a newly created branch (how would we name it?)
3. Checkout in [detached HEAD mode](https://www.git-tower.com/learn/git/faq/detached-head-when-checkout-commit) (this is what I asked above)

**Comment by mikemicmike** [comment]
Ahh i see what you mean. I understand this isn't a git option normally. I think checking out in detached head would probably confuse people a lot more. Wouldn't the simplest solution be to checkout to the local branch and then do an automatic pull? VsCode already seems to do this because if we use the option to create a new branch from a remote branch then it picks up that remote branch just fine (i dont know exactly how this works obviously).

I just feel like what would feel the most natural to me using this would be that if i swap to a branch with my remote label in front of it, git would automatically perform a pull request thus giving me the latest version available on that server. Yes this would automatically override your local branch with a pull request as well but I dont see a real inconvenience in doing this. 

Basically, I think indeed option 1 is the best course of action but adding an automatic pull request if this was a remote branch being checked out. 

Thanks!

**Comment by joaomoreno** [comment]
> Wouldn't the simplest solution be to checkout to the local branch and then do an automatic pull?

Doing an automatic pull as part of a `Checkout` command would even be more confusing, wouldn't you say?

I think we'll keep the current behavior for now, thanks for the discussion though!
